torch/csrc/autograd/input_buffer.cpp

```diff
@@ -129,96 +149,147 @@ static void accumulate(
-        opt_accumulate_stream = guard.getDefaultStream(*device);
-      }
-      if (opt_sync_stream && (opt_accumulate_stream != opt_sync_stream)) {
-        // (3b), (4b)
-        c10::OptionalDeviceGuard device_guard{opt_sync_stream->device()};
-        auto event = c10::Event{device_type};
-        event.record(*opt_sync_stream);
-        opt_accumulate_stream->wait(event);
-        const auto guard = c10::impl::VirtualGuardImpl(device_type);
-        record_stream_any_impl(var, *opt_accumulate_stream);
-      }
+      c10::OptionalDeviceGuard device_guard{device};
+      accumulate(buffer, pos, std::move(var));
     }
+    return;
   }
 
-  auto& old_var = buffer[pos];
-  if (!old_var.defined()) {
+  // Handle the case where var is on accelerator but producer node has no
+  // canonical stream, e.g. this can happen if forward is DtoH
+  const std::optional<c10::Stream>& opt_producer_stream =
+      (opt_producer_stream_.has_value()
+           ? opt_producer_stream_
+           : std::optional<c10::Stream>(
+                 at::accelerator::getCurrentStream(device->index())));
+
+  TORCH_INTERNAL_ASSERT(opt_consumer_stream && opt_producer_stream);
+
+  //
+  // [ Single producer case ]
+  //
+  if (num_dependencies == 1) {
+    mb_wait_stream(opt_consumer_stream, opt_producer_stream, device_type);
+    mb_record_stream_any_impl(
+        var, *opt_consumer_stream, /*prev_stream=*/*opt_producer_stream);
+    TORCH_INTERNAL_ASSERT(current_index == 0);
+    TORCH_INTERNAL_ASSERT(!buffer[pos].defined());
     buffer[pos] = std::move(var);
-  } else {
-    if (opt_accumulate_stream) {
-      c10::OptionalStreamGuard stream_guard{opt_accumulate_stream};
-      accumulate(buffer, pos, std::move(var));
+    return;
+  }
+  //
+  // [ Multiple producer case ]
+  //
+  // First producer
+  // ==============
+  //
+  // A. Determine the accumulation stream:
+  //    case 1) var's device matches consumer node's canonical device
+  //            (The producer node's canonical device may or may not match)
+  //            -> accumulator stream = consumer stream
+  //    case 2) var's device only matches producer node's canonical device
+  //            -> accumulator stream = producer stream
+  //    case 3) var device matches neither
+  //            -> accumulator stream = var device's current stream
+  // B. Record stream (var will be used on the accum stream)
+  // C. Because we are the first producer, there's no accumulation necessary.
+  //    Just move var into the buffer.
+  //
+  // Nth producer
+  // ============
+  //
+  // A. The accumulation stream (determined at "first producer") waits for
+  //    the new producer stream
+  // B. Record stream (var will be used on the accum stream)
+  // C. Accumulate var into buffer
+  // D. (Cases 2 and 3 only) If we are the last producer, we need to have
+  //    consumer wait for the accumulation stream. For case 1 this is
+  //    not necessary because the consumer stream is the accumulation stream.
```

Comment:
You have special casing to skip same stream sync so this special case is not needed and it will just be a noop right?
Just to remove a conditional from the code below